### PR TITLE
Use filters for changing human height

### PR DESCRIPTION
**Pull request by jacklecroy**
## Changelog

Tau Ceti Station added height preferences. https://github.com/TauCetiStation/TauCetiClassic/pull/8179 This can be added in a later change.

Six height levels are available, and one is used for dwarfism now.

:cl: That REALLY Good Soda Flavor
code: Changed how dwarf appearance is created.
admin: Admins can adjust a human mob's height in VV menu.
/:cl:

### CODE REVIEW

**Comment by carshalash** [comment]
Aww, I wanted manlets of every species.

**Comment by jacklecroy** [comment]
It should work for any species including monkey.

**Comment by theta060** [comment]
time to become short just so its harder to click on me (and to mirror myself ingame)

**Comment by tralezab** [comment]
image
"mom said it was my turn to use the insulated gloves"

**Comment by axietheaxolotl** [comment]
I can SEE the mixels in the screenshot

**Comment by Fikou** [comment]
the tweak label does not exist anymore

**Review by Fikou — COMMENTED** [review]
these break any sprite that would go above 32x32 (including huds), do you have any solution to that?
image
image

**Comment by jacklecroy** [comment]
> these break any sprite that would go above 32x32 (including huds), do you have any solution to that? image image

I will look into increasing the size of the filter and see if that works. I was having trouble adding a filter to mutable appearances, but I will look into that and then the HUD could be ignored.

**Comment by TheBonded** [comment]
image
why do they not have legs

**Comment by jacklecroy** [comment]
I can't figure out any way to make the sprite not cut off at 32px.

File: code/__DEFINES/mobs.dm [hunk 1]
```diff
@@ -562,6 +562,13 @@
 #define SIGN_TRAIT_BLOCKED 3
 #define SIGN_CUFFED 4
 
+/// Height defines
+#define CARBONHEIGHT_SHORTEST "shortest height"
+#define CARBONHEIGHT_SHORT    "short height"
+#define CARBONHEIGHT_MEDIUM   "medium height"
+#define CARBONHEIGHT_TALL     "tall height"
+#define CARBONHEIGHT_TALLEST  "tallest height"
```
Comment (Mothblocks) [inline code/__DEFINES/mobs.dm:570]: No alignment indentation. https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#tabs-not-spaces

Also, rename to CARBON_HEIGHT, two words.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Why has filters moved to datums? That does not make any sense, filters are an atom-level concept.

**Comment by jacklecroy** [comment]
> Why has filters moved to datums? That does not make any sense, filters are an atom-level concept.

To apply them on an image or mutable appearance.

**Comment by tralezab** [comment]
> I can't figure out any way to make the sprite not cut off at 32px.

This is a really big killer for the pr, having height cut off any overlays on a mob that is entirely built out of overlays that currently has items that extend beyond the 32x32 bounds is just too much. If there's a solution I'm down but that's pretty big killer for me since we need to stop supporting bigge (tm) items and overlays, which can currently be done

**Comment by DarthSidiousPalpatine** [comment]
I am the author of the original PR on the Tau Ceti build. You just missed the one thing I did.

Make two procs, one will apply filter on image regarding the height parameter.
Second will apply pixel_y offset on image regarding the height parameter.

So, when you assemble your body (torso, legs, hands) + body uniform/clothes - you apply filters to them via the first proc.
Then, when you assemble head, masks, inhand items, helmets, tails, hair and etc - you have to apply pixel_y offset via the second proc.
Also, apply appearance flag "KEEP_APART " to your hud images or something like that.

Sorry, I'm too lazy to puzzle out your build and exactly how you assemble your human sprite, so, I just can help with advice.

**Comment by jacklecroy** [comment]
> I am the author of the original PR on the Tau Ceti build. You just missed the one thing I did.
> 
> Make two procs, one will apply filter on image depending the height parameter. Second will apply pixel_y offset on image depending the height parameter.
> 
> So, when you assemble your body (torso, legs, hands) + body uniform/clothes - you apply filters to them via the first proc. Then, when you assemble head, masks, inhand items, helmets, tails, hair and etc - you have to apply pixel_y offset via the second proc. Apply filters and offset directly to the separate overlay images before they combined into a human. Also, apply appearance flag "KEEP_APART " to your hud images or something like that.
> 
> Sorry, I'm too lazy to puzzle out your build and exactly how you assemble your human sprite, so, I just can help with advice.
> 
> The REAL problem with that is, a photos are kinda "broken", just because GetFlatIcon works just by picking an actual object's icon, without any filters that are "client-side". So, you will get strange and weird hair that floats above your human or pushed in your head, weird offsetting and etc. I heard BYOND devs want to refactor how GetFlatIcon or something like that will work, but I'm trying to solve this with completely new and separate photography mechanic that will work with images instead of icons, but its kinda scetchy now. Nevermind, its a small issue.
> 
> Sorry for my English, its kinda ugly and bad sometimes.

I will look into that! I think tg has a different method for GetFlatIcon. That function is notoriously bad, so I don't think it will matter much.

**Pull request closed without merging**